PYIC-6236: Make ProcessAsyncCriCredentialHandler inclusive for DCMAW

MikeCollingwood opened the pull request:
## Proposed changes

### What changed

- Use state to index userId's cri response instead of criId.
- Make sending audit events conditional only for F2F cri.
- Remove criId from expected queue response dto.

### Why did it change

- We need to find out the criId instead of hardcoding F2F, as we are trying to be inclusive to DCMAW async VCs.
- Cri Id isn't sent via queue message, but state is.
- Audit events are only expected for F2F at the moment.

### Issue tracking

- [PYIC-5039](https://govukverify.atlassian.net/browse/PYIC-5039)

### Notes

- Tested it and it works: DCMAW async VC gets stored & no audit events
- Still want to check assumptions (see comments)
- CRI oauth configs to have signing key

[PYIC-5039]: https://govukverify.atlassian.net/browse/PYIC-5039?atlOrigin=eyJpIjoiNWRkNTljNzYxNjVmNDY3MDlhMDU5Y2ZhYzA5YTRkZjUiLCJwIjoiZ2l0aHViLWNvbS1KU1cifQ

MikeCollingwood committed:
PYIC-5039: Make ProcessAsyncCriCredentialHandler inclusive for DCMAW
- Use state to index userId's cri response instead of criId.
- Make sending audit events conditional only for F2F cri.
- Remove criId from expected queue response dto.

MikeCollingwood commented on lambdas/process-async-cri-credential/src/main/java/uk/gov/di/ipv/core/processasynccricredential/ProcessAsyncCriCredentialHandler.java:
```diff
@@ -167,44 +168,49 @@ private void processErrorAsyncCriResponse(ErrorAsyncCriResponse errorAsyncCriRes
+                        .with(LOG_ERROR_CODE.getFieldName(), errorAsyncCriResponse.getError()));
 
-        sendIpvVcErrorAuditEvent(errorAsyncCriResponse);
+        if (criResponseItem.isPresent()
```
This will only be called when there is a cri response, as we need it to check for F2F cri Id.

MikeCollingwood committed:
PYIC-5039: Update unit tests

MikeCollingwood committed:
PYIC-5039: Fix SonarCloud issues

MikeCollingwood commented on lambdas/process-async-cri-credential/src/main/java/uk/gov/di/ipv/core/processasynccricredential/ProcessAsyncCriCredentialHandler.java:
```diff
@@ -224,51 +228,47 @@ private void processSuccessAsyncCriResponse(SuccessAsyncCriResponse successAsync
-                        auditEventUser,
-                        getExtensionsForAudit(verifiableCredential, isSuccessful));
-        auditService.sendAuditEvent(auditEvent);
+        if (Cri.F2F.equals(cri)) {
```
It might be nicer to put this logic in the parent method but it makes it too complex according the SonarCloud

blakeyp replied:
I would agree but I think this is ok for now as we do have a follow-up to handle audit events for both CRIs (possibly with new generic events)

blakeyp commented on lambdas/process-async-cri-credential/src/main/java/uk/gov/di/ipv/core/processasynccricredential/ProcessAsyncCriCredentialHandler.java:
```diff
@@ -167,44 +168,47 @@ private void processErrorAsyncCriResponse(ErrorAsyncCriResponse errorAsyncCriRes
+        Optional<CriResponseItem> criResponseItem = getCriResponseItem(userId, state);
+
+        if (criResponseItem.isEmpty()) {
+            LOGGER.error(LogHelper.buildLogMessage("No response item found"));
```
maybe 'No response item found given user id and state'? as this is effectively replacing the state validation

blakeyp commented on lambdas/process-async-cri-credential/src/main/java/uk/gov/di/ipv/core/processasynccricredential/ProcessAsyncCriCredentialHandler.java:
```diff
@@ -167,44 +168,47 @@ private void processErrorAsyncCriResponse(ErrorAsyncCriResponse errorAsyncCriRes
+        }
 
-        configService.setFeatureSet(criResponseItem.getFeatureSet());
+        var cri = Cri.fromId(criResponseItem.get().getCredentialIssuer());
```
Where do we actually set the CRI id on the response item? I seem to remember we just assumed it's F2F, but this could become a problem if we need to retrieve and distinguish pending records in check existing identity 🤔

MikeCollingwood replied:
It's set by (the entire cri response created by) the `call-dcmaw-async-cri` which produces a "pending" CRI response before you fill in details to get the VC which then appears on the queue. In this handler there is a section of code:

```java
criStoringService.recordCriResponse(
    request, DCMAW_ASYNC, oauthState, clientOAuthSessionItem, featureSets);
```

Which saves the cri response with `dcmawAsync` as the cri id. So it should be ok as we have also seen these cri responses in DynamoDb during testing.

We _did_ use to assume criId=F2F when creating the `(success/error)AsyncCriResponse` object but you can see we have altogether removed that attribute now as it is really just a dto for the queue response which doesn't include a cri id.

blakeyp commented on lambdas/process-async-cri-credential/src/main/java/uk/gov/di/ipv/core/processasynccricredential/ProcessAsyncCriCredentialHandler.java:
```diff
@@ -224,51 +228,47 @@ private void processSuccessAsyncCriResponse(SuccessAsyncCriResponse successAsync
+        }
     }
 
     private void sendIpvVcErrorAuditEvent(ErrorAsyncCriResponse errorAsyncCriResponse) {
```
to be consistent it might be clearer to do the same here (ie. move the conditional inside the method), for now?

MikeCollingwood committed:
PYIC-5039:
- Refactor cri check into sendIpvVcErrorAuditEvent
- Update empty criResponseItem message
